### PR TITLE
Resizing Layer

### PR DESCRIPTION
**Pull request by koyykdy**
* Rescaling Preprocessing Layer Co-authored-by:
David Kim (@koyykdy) <dok098@ucsd.edu>
Brian Zheng (@brianzheng123) <brianzheng345@gmail.com>

* Version issues resolved

* linting and PR issues resolved
Co-authored-by: Adam Lang (@AdamLang96) <adamglang96@gmail.com>
Co-authored-by: (@Brianzheng123) <brianzheng345@gmail.com>

* initial implementation of image preprocessing: resizing layer, and associated unit tests. Comments and refactoring for image scaling layer

* License refactoring in computeOutputShape for image resizing layer

* Unit tests for image resizing preprocessing layer expanded and refactored

* refactored unit tests for resizing layer

* Preprocessing-Resizing layer unit test expansion and refactoring.
Co-authored-by: Adam Lang (@AdamLang96) <adamglang96@gmail.com>

* cleaning up commit diffs

Co-authored-by: AdamLang96 <45542095+AdamLang96@users.noreply.github.com>

To see the logs from the Cloud Build CI, please join either our [discussion](https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs) or [announcement](https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs-announce) mailing list.

- - -
This change is [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/6858)

### CODE REVIEW

**Comment by koyykdy** [comment]
IMPORT NOTE:
The lower-level op implementation of image resizing-nearest neighbor in TensorFlow.js differs from the implementation of the comparable op in Keras-Python.

While the Python version of the op function always selects the bottom right cell of the sub-matrix to be used as the representative value of that region in the downscaled matrix, the JavaScript implementation defaults to the top left cell of the sub-matrix, and then preferentially shifts to the right side of the sub-matrix in all sub-matrices past the lateral halfway point ( calculated by floor((length-1)/2) ), and the bottom side of the sub-matrix in all sub-matrices past the vertical halfway point, when considering the top-left side of the parent matrix as the origin.

This causes a slight variation in the output values from nearest neighbor downscaling between the Python and JavaScript versions of the code as it currently stands, and the unit tests for the resizing layer has been implemented to reflect this difference in op-function behavior.

**Review by mattsoulanille — CHANGES_REQUESTED** [review]
Thanks for the PR! Sorry it took me a while to get to. I have just a few small changes.

File: tfjs-layers/src/layers/preprocessing/image_resizing_test.ts [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+    expectTensorsClose(layerOutputTensor, inputTensor);
+  });
+
+  it('Returns a tensor of the correct dtype', () => {
+    // do a same resizing operation, cheeck tensors dtypes and content
+    const height = 40;
+    const width = 60;
+    const numChannels = 3;
+    const inputTensor: Tensor<Rank.R3> =
+        randomNormal([height, width, numChannels]);
```
[the remark this answers is not in the document]
Comment (koyykdy) [reply tfjs-layers/src/layers/preprocessing/image_resizing_test.ts]: Requested changes have been commit in https://github.com/tensorflow/tfjs/pull/6858/commits/8532991569e1495c78344a4d26ef2fc688ab4715

File: tfjs-layers/src/layers/preprocessing/image_resizing_test.ts [hunk 2]
```diff
@@ -86,8 +86,13 @@ describeMathCPUAndGPU('Resizing Layer', () => {
...
     const width = 32;
-    const numChannels = 3;
-    const inputTensor = randomNormal([height, width, numChannels]);
+    const numChannels = 1;
+    const rangeArr = [...Array(height * width).keys()];
+    const inputArr = [];
+    while(rangeArr.length) {
+      inputArr.push(rangeArr.splice(0, width));
+    }
+    const inputTensor = tensor(inputArr, [height, width, numChannels]);
```
Comment (mattsoulanille) [inline tfjs-layers/src/layers/preprocessing/image_resizing_test.ts:95]: Nit: [`tf.range`](https://js.tensorflow.org/api/latest/#range) and [`tf.reshape`](https://js.tensorflow.org/api/latest/#reshape) can make this simpler.

```suggestion
    const rangeTensor = range(0, height * width);
    const inputTensor = reshape(rangeTensor, [height, width, numChannels]);
```

File: tfjs-layers/src/layers/preprocessing/image_resizing.ts [hunk 3]
```diff
@@ -99,7 +91,7 @@ export class Resizing extends Layer {
         return image.resizeNearestNeighbor(
             inputs, size, !this.cropToAspectRatio);
       } else {
-        throw new Error('interpolation error encountered in image_resizing.ts');
+        throw new Error(`mode is ${this.interpolation} but only ${[...INTERPOLATION_METHODS]} are supported`);
```
Comment (mattsoulanille) [inline tfjs-layers/src/layers/preprocessing/image_resizing.ts:94]: Nit: Sorry, my last suggestion forgot to include interpolation.
```suggestion
        throw new Error(`Interpolation is ${this.interpolation} but only ${[...INTERPOLATION_METHODS]} are supported`);
```

**Review by mattsoulanille — APPROVED** [review]
LGTM with some small nits.

@ahmedsabie This PR is ready for your review when you get a chance. Thanks!

File: tfjs-layers/src/layers/preprocessing/image_resizing_test.ts (hunk 2 above)
Comment (mattsoulanille) [reply tfjs-layers/src/layers/preprocessing/image_resizing_test.ts:95]: It looks like this applies to several other tests as well. Sorry I didn't catch this in my first review.

**Pull request closed without merging**